txpipe/rail_pz.py

```diff
@@ -0,0 +1,190 @@
+            output.write(estimator)
+
+
+class PZRailEstimate(PipelineStage):
+    """Run a trained RAIL estimator to estimate PDFs and best-fit redshifts
+
+    We load a redshift Estimator model, typically saved by the PZRailTrain stage,
+    and then load chunks of photometry and run the estimator on it, and save the
+    result.
+
+    RAIL currently returns the PDF and then only the modal z as a point estimate,
+    so we save that.  Previous stages have returned mean or median methods too.
+    We could ask for this in RAIL, or calculate the mean or median ourselves.
+
+    There's currently a slight ambiguity about bin edges vs bin centers that I
+    will follow up on with the RAIL team.
+
+    The training stage is (currently all) serial, but applying the trained
+    model can be done in parallel, so we split into two stages to avoid
+    many processors sitting idle or repeating the same training process.
+
+    """
+    name = "PZRailEstimate"
+
+    inputs = [
+        ("photometry_catalog", HDFFile),
+        ("photoz_trained_model", PickleFile),
+    ]
+
+    outputs = [
+        ("photoz_pdfs", PhotozPDFFile),
+    ]
+
+    config_options = {
+        "chunk_rows": 10000,
+    }
+
+    def run(self):
+        # Importing this means that we can unpickle the relevant class
+        import rail.estimation
+
+        # Load the estimator trained in PZRailTrain
+        with self.open_input("photoz_trained_model", wrapper=True) as f:
+            estimator = f.read()
+
+        # prepare the output data - we will save things to this
+        # as we go along
+        output = self.setup_output_file(estimator)
+
+        # Create the iterator the reads chunks of photometry
+        # The method we use here automatically splits up data when we run in parallel
+        cols = [f"mag_{b}" for b in "ugrizy"] + [f"mag_err_{b}" for b in "ugrizy"]
+        chunk_rows = self.config["chunk_rows"]
+        it = self.iterate_hdf("photometry_catalog", "photometry", cols, chunk_rows)
+
+        # Loop through the chunks of data
+        for s, e, data in it:
+            # Rename things so col names match what RAIL expects
+            self.rename_columns(data)
+
+            # Run the pre-trained estimator
+            pz_data = estimator.estimate(data)
+
+            # Save the results
+            self.write_output_chunk(output, s, e, pz_data)
+
+    def rename_columns(self, data):
+        # RAIL expects the magnitudes and errors to have
+        # the suffix _lsst, which we add here
+        for band in "ugrizy":
+            data[f"mag_{band}_lsst"] = data[f"mag_{band}"]
+            data[f"mag_err_{band}_lsst"] = data[f"mag_err_{band}"]
+
+    def setup_output_file(self, estimator):
+        # Briefly check the size of the catalog so we know how much
+        # space to reserve in the output
+        with self.open_input("photometry_catalog") as f:
+            nobj = f["photometry/ra"].size
```

Comment:
what format needs `f` be for `"photometry/ra"` to be legit?

Reply:
HDF5.   Which it always is here - this photometry input file here is the one we use within TXPipe, which we first extract from GCR to disc in an earlier stage (in `input_cats.py`).

Reply:
ok I see, we stopped short from defining a write interface for pz products, but we are probably going to have to do it, to ensure consistency..... so this will need to be revised anyway

Reply:
I have a parallel issue to switch to using `qp` for data storage.  Once I have this bit worked out I'll look into that.

Reply:
There is already a RAIL branch with a quick and dirty switch to qp storage. 
https://github.com/LSSTDESC/RAIL/tree/issue/33/qp-output
However, as mentioned in the associated issue 
https://github.com/LSSTDESC/RAIL/issues/33
there is currently an awkward hack where we write to multiple astropy table chunks as an interim step and then join them at the end in an inefficient way (note that RAIL has its own data chunking separate from TXPipe).
If there's a way to write astropy tables with parallel writes I'd be happy to switch this out. Or, we may want to talk about switching qp storage to just straight hdf5 rather than astropy tables, since parallel write is pretty straightforward for that format.